### ISSUE TITLE
Understanding exports_sources and CMake's add_subdirectory

### ISSUE THREAD

**Issue by kobi-ca** [conan-io/conan#9906]
Hello Kings of the swamp,

My understanding is that when I have something like:
```
    exports_sources = "CMakeLists.txt", "src/*", "tools/*", "tests/*", "include/*"
```

`conan create` would take all of these and place it in the `.conan/...` (aka the recipe).
I also realized, though I'm not sure, that if code in `tests/CMakeLists.txt` need some package to link with, e.g. `boost` but the **rest** of the tree does not need it, since I have it in `exports_sources` it is as if the _entire package_ needs `boost`.

Now, I also figured out that `add_subdirectory()` need to match the above `exports_sources` list.
In other words, if I have `add_subdirectory(tests)` and I remove `tests/*` from the above list, I have an error:
```
CMake Error at CMakeLists.txt:43 (add_subdirectory):
  add_subdirectory given source "tests" which is not an existing directory.
...
ERROR: foo/1.1.1: Error in build() method, line 41
	cmake.configure()
	ConanException: Error 1 while executing cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=..."
```

But I don't want `tests/*`. I don't need it in the recipe. I don't need the dependencies.
How do I remove `tests/*` without the above error. Surely there is a way.

Thanks!
Kobi

**Comment by memsharded** [comment]
> I also realized, though I'm not sure, that if code in tests/CMakeLists.txt need some package to link with, e.g. boost but the rest of the tree does not need it, since I have it in exports_sources it is as if the entire package needs boost.

if you wanted to maintain "tests" and build and run them when you build the packge, this might be a case for ``build_requires``. This dependency of boost is used only for running the tests, but as you said, boost will not be linked or be part of the final packaged library or executable in any way. In this sense, boost is like "catch2", "gtest", or "cmake", something that is exclusively used at build/test time.

If, on the other hand as you say, you want to remove the tests completely from the package build and creation, then the approach would be:
- Do not exports the ``tests/*`` into the recipe.
- Protect your ``CMakeLists.txt`` with some conditional var like BUILD_TESTS, so it doesn't ``add_subdirectory(tests)`` if it is not enable.
- Disable the ``BUILD_TESTS`` var in the recipe (for example, via ``CMakeToolchain``)

**Comment by kobi-ca** [comment]
Thanks @memsharded trying it now. Looks promising. Still building.
I added:
```
    build_requires = ["boost/1.69.0"]
```

and:
```
        tc.variables["FOO_BUILD_TESTS"] = "OFF"
```

and 
```
if(FOO_BUILD_TESTS)
    message("Building FOO tests")
    find_package(Boost REQUIRED)
    add_subdirectory(tests)
endif()
```

I needed to place `find_package(Boost REQUIRED)` inside the `if` since otherwise there is an error of `BZip2` not found. not sure yet why but I'll get back to it later.

BTW, for some reason, building is never parallel (i.e. like there is no `-j`)

**Comment by kobi-ca** [comment]
OK. It worked. Thank you so much as usual. Take a look at my message above and see if you have any comments. otherwise, we can close it. thanks!

**Comment by kobi-ca** [comment]
OK. here is what I found:

this works when running conan create .`
```
requires = ["fmt/8.0.1", "boost/1.69.0"]
```
But, as soon as I try to put Boost in build_requires, it forces me (otherwise there are errors) to explicitly bring in all the deps:
```
    requires = ["fmt/8.0.1"]
    build_requires = ["boost/1.69.0", "bzip2/1.0.8", "zlib/1.2.11", "libbacktrace/cci.20210118"]
```
Needed to add one by one and eventually satisfy the whole dep-tree:
https://conan.io/center/boost?version=1.69.0&tab=dependencies

what am I missing here?

Thanks!

**Comment by memsharded** [comment]
I think it might be related to the nature of ``build_requires``, not intended to be used via ``find_package(), and this also depends on if you are using the "build" and "host" profiles. The direction should be:

- Put ``boost`` as ``self.requires("boost...", force_host_context=True)`` inside ``requirements()``, this indicates this dep is in the "host" context.
- That should handle the transitive dependencies too (using the new ``CMakeDeps`` integrations.
- Recommended to use the ``--profile:build`` explicitly, to start opting-in in what will be the default behavior in 2.0 and in ConanCenter soon too.

**Comment by kobi-ca** [comment]
thank you @memsharded !!
